lib/inject.ex

```diff
@@ -8,19 +8,21 @@ defmodule Inject do
     :ok
   end
 
-  def i(mod) do
-    inject(mod)
+  defmacro inject(mod) do
+    quote bind_quoted: [mod: mod] do
+      if Application.get_env(:inject, :disabled) do
+        mod
+      else
+        Inject.Registry
+        |> Registry.lookup(mod)
+        |> Enum.reverse()
+        |> find_override() || mod
+      end
+    end
   end
```

Comment:
The best way to visualise macros (well, for me at least) is to think of them as copy & paste.
With macro defined like this we would get this code:

```ex
def fancy do
  i(Dep).fun()
end
```

replaced with

```ex
def fancy do
  (
      if Application.get_env(:inject, :disabled) do
         mod
       else
         Inject.Registry
         |> Registry.lookup(mod)
         |> Enum.reverse()
         |> find_override() || mod
       end
  ).fun()
end
```

This would still call the `Application.get_env` every time `fancy/0` is called.

What we want instead, is to evaluate the `Application.get_env` call in compile time, inside macro:

```ex
defmacro inject(mod) do
  # run stuff here, during compilation
  if letsdothis do
    quote(do: injectstuff)
  else
    quote(do: nah)
end
```

This way the body of `fancy/0` would be either `(injectsutff).fun()` or `(nah.fun())`.

Now, what should `injectsutff` and `nah` be is not for me to decide ;) 
But I'll give you one more hint - it's usually a good practice to put the least amount of code inside quote as possible. If you can replace some lines with a function call and just call that function inside quote it should be done.

Can't wait to see the v2! 🎸

Reply:
I took a stab at V2. I just amended the commit; should have just squashed down before merging so you could see the change history but oh well, at least the change is small! I also asked a separate question about mod which you can see in the diff.

By the way, the test in `test/inject_configuration_test.exs` now fails, since changing the environment at runtime won't change the compiled code. Is there a way to recompile the code just for a test, or should we just remove this check? For what it's worth, I don't know if it's possible to do this from a quick google test.